src/databricks/labs/ucx/source_code/python_libraries.py

```diff
@@ -45,18 +60,64 @@ def _temporary_virtual_environment(self):
 
     def _install_library(self, path_lookup: PathLookup, library: Path) -> list[DependencyProblem]:
         """Pip install library and augment path look-up to resolve the library at import"""
-        venv = self._temporary_virtual_environment
-        path_lookup.append_path(venv)
-        # resolve relative pip installs from notebooks: %pip install ../../foo.whl
+        path_lookup.append_path(self._temporary_virtual_environment)
+
+        # Resolve relative pip installs from notebooks: %pip install ../../foo.whl
         maybe_library = path_lookup.resolve(library)
         if maybe_library is not None:
             library = maybe_library
-        return_code, stdout, stderr = self._runner(f"pip install {library} -t {venv}")
+
+        if library.suffix == ".egg":
+            return self._install_egg(library)
+        return self._install_pip(library)
+
+    def _install_pip(self, library: Path) -> list[DependencyProblem]:
+        return_code, stdout, stderr = self._runner(f"pip install {library} -t {self._temporary_virtual_environment}")
         logger.debug(f"pip output:\n{stdout}\n{stderr}")
         if return_code != 0:
             problem = DependencyProblem("library-install-failed", f"Failed to install {library}: {stderr}")
             return [problem]
         return []
 
+    def _install_egg(self, library: Path) -> list[DependencyProblem]:
+        """Install egss using easy_install.
+
+        Sources:
+            See easy_install in setuptools.command.easy_install
+        """
+        verbosity = "--verbose" if is_in_debug() else "--quiet"
+        easy_install_arguments = [
+            "easy_install",
+            verbosity,
+            "--always-unzip",
+            "--install-dir",
+            self._temporary_virtual_environment.as_posix(),
+            library.as_posix(),
+        ]
+        has_setup = True  # setup might not be present
+        try:
+            # Mypy can't analyze setuptools due to missing type hints
+            from setuptools import setup  # type: ignore
+        except ImportError:
+            try:
+                from distutils.core import setup  # pylint: disable=deprecated-module
+            except ImportError:
+                logger.warning("Could not import setup.")
+                has_setup = False
+        if has_setup:
+            try:
+                setup(script_args=easy_install_arguments)
+                return []
+            except (SystemExit, ImportError, ValueError):
+                pass
+        try:
+            # Not a "real" install, though, the best effort to still lint eggs without dependencies
+            with zipfile.ZipFile(library, "r") as zip_ref:
+                zip_ref.extractall(self._temporary_virtual_environment)
+        except zipfile.BadZipfile as e:
+            problem = DependencyProblem("library-install-failed", f"Failed to install {library}: {e}")
+            return [problem]
+        return []
```

Comment:
```suggestion
        def _get_setup():
            try: 
                # Mypy can't analyze setuptools due to missing type hints
                from setuptools import setup  # type: ignore
                return setup
             except ImportError:
                try:
                   from distutils.core import setup  # pylint: disable=deprecated-module
                   return setup
                except ImportError:
                   return None
        setup = _get_setup()
        if setup:
           try:
                setup(script_args=easy_install_arguments)
                return []
            except (SystemExit, ValueError) as e:
                problem = DependencyProblem("library-install-failed", f"Failed to install {library}: {e}")
                return [problem]
        try:
            # Not a "real" install, though, the best effort to still lint eggs without dependencies
            with zipfile.ZipFile(library, "r") as zip_ref:
                zip_ref.extractall(self._temporary_virtual_environment)
        except zipfile.BadZipfile as e:
            problem = DependencyProblem("library-install-failed", f"Failed to install {library}: {e}")
            return [problem]
        return []
```
or even split it to more confined methods. we'll have bugs here anyway and small methods are easier to fix :)